File: src/nodetool/models/message.py
import json
from datetime import datetime
from typing import Any

from pydantic import field_validator

from nodetool.metadata.types import MessageContent, MessageFile, Provider, ToolCall
from nodetool.models.base_model import DBField, DBModel, create_time_ordered_uuid
from nodetool.models.condition_builder import Field


class Message(DBModel):

    @field_validator("tools", "collections", mode="before")
    @classmethod
    def _deserialize_str_list(cls, v: Any) -> Any:
        """Handle list fields arriving as JSON strings from SQLite."""
        if isinstance(v, str):
            try:
                parsed = json.loads(v)
                if isinstance(parsed, list):
                    return parsed
            except (json.JSONDecodeError, TypeError):
                pass
        return v

    @field_validator("tool_calls", "input_files", "output_files", mode="before")
    @classmethod
    def _deserialize_obj_list(cls, v: Any) -> Any:
        """Handle complex list fields arriving as JSON strings from SQLite."""
        if isinstance(v, str):
            try:
                parsed = json.loads(v)
                if isinstance(parsed, list):
                    return parsed
            except (json.JSONDecodeError, TypeError):
                pass
        return v
    @classmethod
    def get_table_schema(cls):
        return {
            "table_name": "nodetool_messages",
        }

    id: str = DBField()
    user_id: str = DBField(default="")
    workflow_id: str | None = DBField(default=None)
    graph: dict | None = DBField(default=None)
    thread_id: str | None = DBField(default=None)
    tools: list[str] | None = DBField(default=None)
    tool_call_id: str | None = DBField(default=None)
    role: str | None = DBField(default=None)
    name: str | None = DBField(default=None)
    content: str | dict[str, Any] | list[MessageContent] | None = DBField(default=None)
    tool_calls: list[ToolCall] | None = DBField(default=None)
    collections: list[str] | None = DBField(default=None)
    input_files: list[MessageFile] | None = DBField(default=None)
    output_files: list[MessageFile] | None = DBField(default=None)
    created_at: datetime | None = DBField(default=None)
    provider: Provider | None = DBField(default=None)
    model: str | None = DBField(default=None)
    cost: float | None = DBField(default=None)
    agent_mode: bool | None = DBField(default=None)
    help_mode: bool | None = DBField(default=None)
    agent_execution_id: str | None = DBField(default=None)
    execution_event_type: str | None = DBField(default=None)
    workflow_target: str | None = DBField(default=None)

    @classmethod
    async def create(cls, thread_id: str, user_id: str, **kwargs) -> "Message":  # type: ignore[override]
        if ("instructions" in kwargs and "content" not in kwargs) or (
            "instructions" in kwargs and kwargs.get("content") is None
        ):
            kwargs["content"] = kwargs.pop("instructions")
        return await super().create(
            id=kwargs.get("id") is None and create_time_ordered_uuid(),
            thread_id=thread_id,
            user_id=user_id,
            **kwargs,
        )

    @classmethod
    async def paginate(
        cls,
        thread_id: str | None = None,
        limit: int = 100,
        start_key: str | None = None,
        reverse: bool = False,
    ):
        return await cls.query(
            condition=Field("thread_id").equals(thread_id).and_(Field("id").greater_than(start_key or "")),
            limit=limit,
            reverse=reverse,
        )

File: ts/packages/models/src/message.ts
/**
 * Message model – conversation messages with tool call support.
 *
 * Port of Python's `nodetool.models.message`.
 */

import type { TableSchema } from "./database-adapter.js";
import type { Row } from "./database-adapter.js";
import {
  DBModel,
  createTimeOrderedUuid,
  type IndexSpec,
  type ModelClass,
} from "./base-model.js";
import { field } from "./condition-builder.js";

// ── Schema ───────────────────────────────────────────────────────────

const MESSAGE_SCHEMA: TableSchema = {
  table_name: "messages",
  primary_key: "id",
  columns: {
    id: { type: "string" },
    user_id: { type: "string" },
    thread_id: { type: "string" },
    role: { type: "string" },
    content: { type: "string", optional: true },
    tool_calls: { type: "json", optional: true },
    input_files: { type: "json", optional: true },
    output_files: { type: "json", optional: true },
    provider: { type: "string", optional: true },
    model: { type: "string", optional: true },
    cost: { type: "number", optional: true },
    workflow_id: { type: "string", optional: true },
    created_at: { type: "datetime" },
    updated_at: { type: "datetime" },
  },
};

const MESSAGE_INDEXES: IndexSpec[] = [
  {
    name: "idx_messages_thread_id",
    columns: ["thread_id"],
    unique: false,
  },
];

// ── Model ────────────────────────────────────────────────────────────

export class Message extends DBModel {
  static override schema = MESSAGE_SCHEMA;
  static override indexes = MESSAGE_INDEXES;

  declare id: string;
  declare user_id: string;
  declare thread_id: string;
  declare role: string;
  declare content: string | null;
  declare tool_calls: unknown[] | null;
  declare input_files: unknown[] | null;
  declare output_files: unknown[] | null;
  declare provider: string | null;
  declare model: string | null;
  declare cost: number | null;
  declare workflow_id: string | null;
  declare created_at: string;
  declare updated_at: string;

  constructor(data: Row) {
    super(data);
    const now = new Date().toISOString();
    this.id ??= createTimeOrderedUuid();
    this.role ??= "user";
    this.content ??= null;
    this.tool_calls ??= null;
    this.input_files ??= null;
    this.output_files ??= null;
    this.provider ??= null;
    this.model ??= null;
    this.cost ??= null;
    this.workflow_id ??= null;
    this.created_at ??= now;
    this.updated_at ??= now;
  }

  override beforeSave(): void {
    this.updated_at = new Date().toISOString();
  }

  // ── Static queries ───────────────────────────────────────────────

  /** Paginate messages in a thread. */
  static async paginate(
    threadId: string,
    opts: { limit?: number } = {},
  ): Promise<[Message[], string]> {
    const { limit = 50 } = opts;
    const cond = field("thread_id").equals(threadId);

    return (Message as unknown as ModelClass<Message>).query({
      condition: cond,
      orderBy: "created_at",
      limit,
    });
  }
}
